ddtrace/tracer/rules_sampler.go

```diff
@@ -0,0 +1,570 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016 Datadog, Inc.
+
+package tracer
+
+import (
+	"encoding/json"
+	"fmt"
+	"math"
+	"os"
+	"regexp"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"golang.org/x/time/rate"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/ext"
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/log"
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/samplernames"
+)
+
+// rulesSampler holds instances of trace sampler and single span sampler, that are configured with the given set of rules.
+// traceRulesSampler samples trace spans based on a user-defined set of rules and might impact sampling decision of the trace.
+// singleSpanRulesSampler samples individual spans based on a separate user-defined set of rules and
+// cannot impact the trace sampling decision.
+type rulesSampler struct {
+	traces *traceRulesSampler
+	spans  *singleSpanRulesSampler
+}
+
+// newRulesSampler configures a *rulesSampler instance using the given set of rules.
+// Rules are split between trace and single span sampling rules according to their type.
+// Such rules are user-defined through environment variable or WithSamplingRules option.
+// Invalid rules or environment variable values are tolerated, by logging warnings and then ignoring them.
+func newRulesSampler(rules []SamplingRule) *rulesSampler {
+	var spanRules, traceRules []SamplingRule
+	for _, rule := range rules {
+		if rule.Type == SamplingRuleSpan {
+			rule.limiter = newSingleSpanRateLimiter(rule.MaxPerSecond)
+			spanRules = append(spanRules, rule)
+		} else {
+			traceRules = append(traceRules, rule)
+		}
+	}
+	return &rulesSampler{
+		traces: newTraceRulesSampler(traceRules),
+		spans:  newSingleSpanRulesSampler(spanRules),
+	}
+}
+
+func (r *rulesSampler) sampleTrace(s *span) bool {
+	return r.traces.apply(s)
+}
+
+func (r *rulesSampler) sampleSpan(s *span) bool {
+	return r.spans.apply(s)
+}
+
+func (r *rulesSampler) HasSpanRules() bool {
+	return r.spans.enabled()
+}
+
+func (r *rulesSampler) TraceRateLimit() (float64, bool) {
+	return r.traces.limit()
+}
+
+// SamplingRule is used for applying sampling rates to spans that match
+// the service name, operation name or both.
+// For basic usage, consider using the helper functions ServiceRule, NameRule, etc.
+// See `SamplingRuleType` for more details.
+type SamplingRule struct {
+	Service      *regexp.Regexp
+	Name         *regexp.Regexp
+	Rate         float64
+	MaxPerSecond float64
+	Type         SamplingRuleType
```

Comment:
Since this is a core part of the sampling API for the user, I think it makes complete sense to document all these fields and how they should be used.